### PR TITLE
Advanced

### PR DESCRIPTION
**Pull request by herbepau**

### CODE REVIEW

File: Breeze.UI/src/app/shared/services/api.service.ts [hunk 1]
```diff
@@ -28,7 +28,7 @@ export class ApiService {
     private headers = new Headers({'Content-Type': 'application/json'});
     private pollingInterval = 3000;
     private bitcoinApiUrl = 'http://localhost:37220/api';
-    private stratisApiUrl = 'http://localhost:37221/api';
+    private stratisApiUrl = 'http://localhost:37221/api' ;
```
Comment (fenix2222) [inline Breeze.UI/src/app/shared/services/api.service.ts]: Why a space?

File: Breeze.UI/src/app/wallet/advanced/advanced.component.css [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    margin-left: 100px;
+    margin-top: 20px;
+}
+
+::ng-deep ng-datepicker input {
+    height: 35px;
+}
+
+.label {
+    font-size:14px;
```
Comment (fenix2222) [inline Breeze.UI/src/app/wallet/advanced/advanced.component.css]: spaces after a colon for consistency

File: Breeze.UI/src/app/wallet/advanced/advanced.component.html [hunk 3]
```diff
@@ -0,0 +1,42 @@
+<div class="mainDiv">
+    <form [formGroup]='icoFormGroup'>
+        <div style="margin-top:20px">
```
Comment (fenix2222) [inline Breeze.UI/src/app/wallet/advanced/advanced.component.html]: As per our discussion please tidy up all inline styles. There seems to be an abuse of those

File: Breeze.UI/src/app/wallet/advanced/advanced.component.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import { LoadingState } from './loadingState';
+import { SerialDisposable } from './serialDisposable';
+
+@Component({
+  selector: 'app-advanced',
+  templateUrl: './advanced.component.html',
+  styleUrls: ['./advanced.component.css']
+})
+export class AdvancedComponent implements OnInit, OnDestroy {
+    private addressCount = "";
```
Comment (fenix2222) [inline Breeze.UI/src/app/wallet/advanced/advanced.component.ts]: single quotes for consistency. My tslint plugin usually complains

File: Breeze.UI/src/app/wallet/advanced/advanced.component.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    private resyncSubs = new SerialDisposable();
+    private addresses: string[] = [];
+    private resyncActioned = false;
+
+    constructor(private advancedService: AdvancedService, private formBuilder: FormBuilder) { 
+        this.setResyncDates();
+    }
+
+    public icoFormGroup: FormGroup;
+    public extPubKey = "";
```
Comment (fenix2222) [inline Breeze.UI/src/app/wallet/advanced/advanced.component.ts]: signle quotes as per the above

File: Breeze.UI/src/app/wallet/advanced/advanced.component.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    resync() {
+        this.resyncActioned = true;
+        const date = new Date(this.resyncDate.year, this.resyncDate.month-1, this.resyncDate.day);
+        this.resyncSubs.disposable = this.advancedService.resyncFromDate(date)
+                                              .monitor(this.resyncLoadingState)
+                                              .subscribe();
+    }
+
+    generateAddresses() {
+        this.addresses = [];
```
Comment (fenix2222) [inline Breeze.UI/src/app/wallet/advanced/advanced.component.ts]: `this.addresses.length = 0` is a much faster way of clearing an array without memory re-assignment

File: Breeze.UI/src/app/wallet/advanced/advanced.service.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        return this.httpClient.get(url).map(x => x.toString());
+    }
+
+    public generateAddresses(count: number): Observable<string[]> {
+        const url = `${this.urlPrefix}unusedaddresses?WalletName=${this.walletName}&AccountName=${this.accountName}&Count=${count}`;
+        return this.httpClient.get(url).map(x => this.processAddresses(x));
+    }
+
+    public resyncFromDate(date: Date): Observable<any> {
+        date = new Date(date.getFullYear(), date.getMonth(), date.getDate()); //<- Strip any time values
```
Comment (fenix2222) [inline Breeze.UI/src/app/wallet/advanced/advanced.service.ts]: Just a question, can we get a null or undefined date passed in? If yes, we should handle it

File: Breeze.UI/src/app/wallet/advanced/feedback/feedback.component.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
+import { Component, Input } from '@angular/core';
+
+@Component({
+  selector: 'app-feedback',
+  templateUrl: './feedback.component.html',
+  styleUrls: ['./feedback.component.css']
+})
+export class FeedbackComponent {
+    private _loading = false;
```
Comment (fenix2222) [inline Breeze.UI/src/app/wallet/advanced/feedback/feedback.component.ts]: Is it possible to use `enum` instead, or maybe limited type, like:
```
private state: ' errored' | 'success' | 'loading' | 'idle';
```
Or something similar. The problem with current approach is that you can potentially corrupt the state and have loading = true and success = true, which is incorrect as you can only be in one state at a time. Unless I am missing something.

File: Breeze.UI/src/app/wallet/dashboard/dashboard.component.ts [hunk 9]
```diff
@@ -96,8 +95,9 @@ export class DashboardComponent implements OnInit {
       .subscribe(
         response => {
           if (response.status >= 200 && response.status < 400) {
-            if (response.json().transactionsHistory.length > 0) {
-              historyResponse = response.json().transactionsHistory;
+            const json = response.json();
+            if (json.transactionsHistory) {
```
Comment (fenix2222) [inline Breeze.UI/src/app/wallet/dashboard/dashboard.component.ts]: What do we do if json.transactionsHistory = false. Should we log it or report it to the user?

File: Breeze.UI/src/app/wallet/status-bar/status-bar.component.ts [hunk 10]
```diff
@@ -27,7 +26,7 @@ export class StatusBarComponent implements OnInit {
   constructor(private apiService: ApiService, private globalService: GlobalService, private genericModalService: ModalService) { }
 
   ngOnInit() {
-    this.startSubscriptions();
+    this.startSubscriptions() ;
```
Comment (fenix2222) [inline Breeze.UI/src/app/wallet/status-bar/status-bar.component.ts]: Why space?

File: Breeze.UI/tsconfig.json [hunk 11]
```diff
@@ -17,7 +17,7 @@
     },
     "types": [
       "node",
-      "jasmine"
+      "jasmine",
```
Comment (fenix2222) [inline Breeze.UI/tsconfig.json]: No need for a comma

File: Breeze.UI/src/app/wallet/advanced/advanced.component.html (hunk 3 above)
Comment (herbepau) [reply Breeze.UI/src/app/wallet/advanced/advanced.component.html]: I have moved styles to classes where that style is used more than once.  If the style is used once then it's overkill.

File: Breeze.UI/src/app/wallet/advanced/advanced.component.css (hunk 2 above)
Comment (herbepau) [reply Breeze.UI/src/app/wallet/advanced/advanced.component.css]: done

File: Breeze.UI/src/app/shared/services/api.service.ts (hunk 1 above)
Comment (herbepau) [reply Breeze.UI/src/app/shared/services/api.service.ts]: done

File: Breeze.UI/src/app/wallet/advanced/advanced.component.ts (hunk 4 above)
Comment (herbepau) [reply Breeze.UI/src/app/wallet/advanced/advanced.component.ts]: done

File: Breeze.UI/src/app/wallet/advanced/advanced.component.ts (hunk 5 above)
Comment (herbepau) [reply Breeze.UI/src/app/wallet/advanced/advanced.component.ts]: done

File: Breeze.UI/src/app/wallet/advanced/advanced.service.ts (hunk 7 above)
Comment (herbepau) [reply Breeze.UI/src/app/wallet/advanced/advanced.service.ts]: Don't think so, but will guard against it

Comment (herbepau) [reply Breeze.UI/src/app/wallet/advanced/advanced.service.ts]: ...actually no, we can't get null - the date is explicitly created by the component.

File: Breeze.UI/src/app/wallet/status-bar/status-bar.component.ts (hunk 10 above)
Comment (herbepau) [reply Breeze.UI/src/app/wallet/status-bar/status-bar.component.ts]: Not something I touched.  Have changed all the same.

File: Breeze.UI/tsconfig.json (hunk 11 above)
Comment (herbepau) [reply Breeze.UI/tsconfig.json]: Not something I touched.  Have changed all the same.

File: Breeze.UI/src/app/wallet/dashboard/dashboard.component.ts (hunk 9 above)
Comment (herbepau) [reply Breeze.UI/src/app/wallet/dashboard/dashboard.component.ts]: Something to discuss.

File: Breeze.UI/src/app/wallet/advanced/feedback/feedback.component.ts (hunk 8 above)
Comment (herbepau) [reply Breeze.UI/src/app/wallet/advanced/feedback/feedback.component.ts]: Can revisit this, but it's such a simple piece of work.  Will leave for now.

File: Breeze.UI/src/app/wallet/advanced/advanced.component.ts (hunk 6 above)
Comment (herbepau) [reply Breeze.UI/src/app/wallet/advanced/advanced.component.ts]: Sure, but performance isn't a concern here, and the reassignment reads better (personally).

Comment (fenix2222) [reply Breeze.UI/src/app/wallet/advanced/advanced.component.ts]: As far as I know it is a bad practice: https://davidwalsh.name/empty-array

Comment (fenix2222) [reply Breeze.UI/src/app/wallet/advanced/advanced.component.ts]: You might get unexpected behavior or memory leaks plus it future proofs it for any new usages of addresses array.

Comment (fenix2222) [reply Breeze.UI/src/app/wallet/advanced/advanced.component.ts]: @herbepau any update?

**Pull request merged**